Merging database extent and ranges with datacube-core

ashoka1234 opened the pull request:
### Reason for this pull request
Extents polygons and bounding boxes are computed for products as a whole as well as for specific time spans. Often list of extents require to be computed for regular intervals which can be yearly, monthly, or daily. Some applications require time bounds as well as spatial bounds product-wide. 

### Proposed changes
Three database tables are proposed to maintain spatial extents, spatial bounds, and time bounds: extent_meta, extent, and ranges. Extent_meta contain meta information regarding spatial extents stored in the extent table. Extent table stores exact spatial boundary of a product for a specific period. The periods are specified with python pandas style timestamp (start) and offset from the start. Offsets are pandas style offset alias strings such as `1M` indicating a month, `1Y` indicating a year etc. The ranges table stores product-wide information such as time bounds and spatial bounds. 

The database-facing retrieval functions are integrated in `datacube.drivers.postgres` module while user-facing retrieval API is located in `ProductResource` class. The schema and migration code are located in `datacube.drivers.postgres` module. The upload scripts are configurable as in `db_extent_cfg.yaml`. The extent and ranges upload script is located in `db_extent_update.py` which rely on a upload API located in `_extent_upload.py`. The uploads are designed to be fast using multiprocessing with task splits across time. Further upload scripts verify what is already done as per information available in the extent_meta table and computes only extents that are not yet computed. Similarly ranges uploads only attempt to see datasets that are not yet seen. So repeated runs of the upload script are fast.  


This pull request replaces #430 

 - [ ] Closes #366 
 - [ ] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

<!--
See https://github.com/blog/2111-issue-and-pull-request-templates for more
information on pull request templates.

-->

ashoka1234 committed:
First commit of extent database code to be integrated with datacube code

ashoka1234 committed:
First commit of extent database code to be integrated with datacube code

ashoka1234 committed:
Placed top level code into functions in the example, added a missing space between arguments

ashoka1234 committed:
Placed top level code into functions in the example, added a missing space between arguments

ashoka1234 committed:
Placed top level code into functions in the example, added a missing space between arguments

ashoka1234 committed:
Placed top level code into functions in the example, added a missing space between arguments

ashoka1234 committed:
First commit of extent database code to be integrated with datacube code

ashoka1234 committed:
First commit of extent database code to be integrated with datacube code

ashoka1234 committed:
First commit of extent database code to be integrated with datacube code

ashoka1234 committed:
First commit of extent database code to be integrated with datacube code

ashoka1234 committed:
Corrected exceptions in get_bounds and get_extent_direct

ashoka1234 committed:
Corrected exceptions in get_bounds and get_extent_direct

ashoka1234 committed:
Corrected exceptions in get_bounds and get_extent_direct

ashoka1234 committed:
Corrected exceptions in get_bounds and get_extent_direct

ashoka1234 committed:
Fixed time out issue relating to sqlalchemy engine objects

ashoka1234 committed:
Fixed time out issue relating to sqlalchemy engine objects

ashoka1234 committed:
Fixed time out issue relating to sqlalchemy engine objects

ashoka1234 committed:
Fixed time out issue relating to sqlalchemy engine objects

ashoka1234 committed:
Refactored the code to improve performance with multiprocessing. Improved representation of empty bounds.

ashoka1234 committed:
Refactored the code to improve performance with multiprocessing. Improved representation of empty bounds.

ashoka1234 committed:
Refactored the code to improve performance with multiprocessing. Improved representation of empty bounds.

ashoka1234 committed:
Refactored the code to improve performance with multiprocessing. Improved representation of empty bounds.

ashoka1234 committed:
Fixed empty extent representations to be consistent with shapely

ashoka1234 committed:
Fixed empty extent representations to be consistent with shapely

ashoka1234 committed:
Fixed empty extent representations to be consistent with shapely

ashoka1234 committed:
Fixed empty extent representations to be consistent with shapely

ashoka1234 committed:
Warning message regarding discarding nanoseconds precision was handled to discard it

ashoka1234 committed:
Warning message regarding discarding nanoseconds precision was handled to discard it

ashoka1234 committed:
Warning message regarding discarding nanoseconds precision was handled to discard it

ashoka1234 committed:
Warning message regarding discarding nanoseconds precision was handled to discard it

ashoka1234 committed:
Added note book examples.

ashoka1234 committed:
Fixed resource leaks. Added note book examples.

ashoka1234 committed:
Fixed resource leaks. Added note book examples.

ashoka1234 committed:
Added note book examples.

ashoka1234 committed:
Schema updated to include constraints and improved normalization
Extent metadata now pre-loaded and the code refactored to reflect it
Uploads to database and retrieval from the database are now in separate files.
Projections added as a parameter to retrieval functions
Retrieval function to intersect extent with a given geobox added

ashoka1234 committed:
Schema updated to include constraints and improved normalization
Extent metadata now pre-loaded and the code refactored to reflect it
Uploads to database and retrieval from the database are now in separate files.
Projections added as a parameter to retrieval functions
Retrieval function to intersect extent with a given geobox added

ashoka1234 committed:
remove _extent_index.py

ashoka1234 committed:
remove _extent_index.py

ashoka1234 committed:
Minor schema (table fields) rename from projection to crs.

ashoka1234 committed:
Minor schema (table fields) rename from projection to crs.

ashoka1234 committed:
Added database maintenance script db_extent_update.py and associated refactorings

ashoka1234 committed:
Added database maintenance script db_extent_update.py and associated refactorings

ashoka1234 committed:
Name change of product_bounds table to ranges. Integrated db-extent schema to datacube.drivers.postgres. database facing functions of db-extent module integrated with PostgresDbAPI class. User facing db_extent function is integrated with datacube.index.ProductResource class.

ashoka1234 committed:
Name change of product_bounds table to ranges. Integrated db-extent schema to datacube.drivers.postgres. database facing functions of db-extent module integrated with PostgresDbAPI class. User facing db_extent function is integrated with datacube.index.ProductResource class.

ashoka1234 committed:
Integrated extent, extent_meta, and ranges tables to datacube schema and migration scripts.

ashoka1234 committed:
Integrated extent, extent_meta, and ranges tables to datacube schema and migration scripts.

ashoka1234 committed:
Integrated extent, extent_meta, and ranges tables to datacube schema and migration scripts.

ashoka1234 committed:
Integrated extent, extent_meta, and ranges tables to datacube schema and migration scripts.

ashoka1234 committed:
refactoring to correct integration errors due to schema merge with datacube-core

ashoka1234 committed:
refactoring to correct integration errors due to schema merge with datacube-core

ashoka1234 committed:
Refactored upload script to be more configurable.

ashoka1234 committed:
Refactored upload script to be more configurable.

ashoka1234 committed:
Refactored store_extent function so that new calls maintain extent_meta table in a consistent state, i.e. all the datasets from start to end are looked at.

ashoka1234 committed:
Refactored store_extent function so that new calls maintain extent_meta table in a consistent state, i.e. all the datasets from start to end are looked at.

ashoka1234 committed:
Refactored code to reflect the removal uuid computes from database retrieval function. Times are casted to dates during search queries.

ashoka1234 committed:
Refactored code to reflect the removal uuid computes from database retrieval function. Times are casted to dates during search queries.

ashoka1234 committed:
Schema updates.

ashoka1234 committed:
Schema updates.

ashoka1234 committed:
Examples reflect merged db-extent code with datacube_core

ashoka1234 committed:
Examples reflect merged db-extent code with datacube_core

ashoka1234 committed:
Added pylint disable too-many-lines to the top of the module

ashoka1234 committed:
Added pylint disable too-many-lines to the top of the module

alexgleith committed:
add in certificate symbolic link

omad commented:
There's a lot of good code in here Ashoka, but there's a few things we need to improve before merging.
- Turn the extents generation tool into an actual tool
- Move the example code out and into documentation
- Include the docs as part of the real documentation
- Add and improve the tests

omad committed:
Move pieces of dbextents into CLI/core code and docs

Flesh out the CLI a bit and integrate it as:
  datacube product_extents update

Delete some old examples.

Add documentation pages to the main documentation.

omad committed:
Move pieces of dbextents into CLI/core code and docs

Flesh out the CLI a bit and integrate it as:
  datacube product_extents update

Delete some old examples.

Add documentation pages to the main documentation.

Kirill888 commented on datacube/scripts/product_extent.py:
```diff
@@ -0,0 +1,73 @@
+_LOG = logging.getLogger('datacube-md-type')
+
+
+@cli.group(name='product_extent', help='Product Extents Tools')
```
Can we just call it `extent`, having `_` in the subcommand feels odd, also we already have just `product` subcommand.

omad replied:
Good idea, we're still working on making it feel integrated.

omad committed:
Some more work on the product extents docs

omad committed:
Geometry object doesn't like None geometries so got to test it. make times from the database are timezone aware

ashoka1234 committed:
comment change

ashoka1234 committed:
unit test user facing functions

ashoka1234 committed:
Units tests on extent functions in ProductResource

ashoka1234 committed:
Merge remote branch 'origin/db_extent' into db_extent

Conflicts:
	datacube/index/_products.py

ashoka1234 committed:
Corrected linting errors

omad committed:
Move product extents tests into the integration tests

They'll be found by Travis and Code Coverage tools in here.

omad committed:
Move product extents tool under `datacube product` cli

omad committed:
Delete old unused product extents code

omad committed:
Try to make overly fussy pylint happy

jeremyh commented:
Thanks Ashoka, it looks good so far. I'm not sure when to add comments as it still seems to be a moving target, and I know Damien already gave you some guidance on integrating it more with the existing code. 

These are a few rough, vague and higher-level notes:

- It still needs some thought to improve how it interacts with drivers: eg. the `extent_upload`is a top-level module, living outside the drivers, but is completely tied to the PostgresDb driver. It's also interacting both with the lowest level api (PostgresDB) and highest level (Datacube) in the same file. We should separate the driver-specific and driver-agnostic parts. Any usage of the high-level api (`Datacube`) prevents it from using these apis due to circular imports (Perhaps replace `dc.find_datasets_lazy` with lower-level `index.datasets.search`?)

- Damien probably already mentioned it, but I'd remove the parts that pass around individual config settings (hostname/password/etc) and use the existing datacube config handling instead. We then also won't need to hard-code dev postgres settings, as people already will have their (dev) settings in `datacube.conf` files.

- The code is creating its own sqlalchemy engine: this might need some cleanup, and if possible use the existing engine setup in PostgresDb (or at least share code with it). It's missing the application name and connection/pool renewal settings and it would be nice not to have to make future connection/alchemy changes in multiple places.

- Currently the pooling/connection handling is disabled, leading to an expensive connection setup (and validation) for almost every query. I ran a test of one month of one product on my laptop and it made ~450 individual connections and ~9400 queries.  Running it a second time (to incrementally update) took significantly longer (I killed it after a few minutes) -- these might be related to #491, though?

- Allow the `extent update` command-line command to take a list of products, and maybe an `--all` option? Most of the time I want to update all of the extents on a system.

- For implementing the tests, there's some existing pytest fixtures that you can use  to test against a real database: have a look at [integration_tests/index/test_search.py](https://github.com/opendatacube/datacube-core/blob/develop/integration_tests/index/test_search.py) as an example that uses them.

- It also might be nice to add a few unit tests for the simple methods with `compute`/`util` in the name. A simple doctest would work on many of them: here's an existing example: (pytest is configured to scan all docstrings for them: [example](https://github.com/opendatacube/datacube-core/blob/d951815a6f8b8bda1e95e7691b00360110358ba5/datacube/model/__init__.py#L681))

- Add a test for updating extents of a product with no datasets, and one with no `time` and/or `lat/lon` fields (eg. satellite_telemetry_data). These are throwing errors.

- It's worth reviewing the API parameter and return types. 
	- Avoid returning sqlalchemy-specific row records: the use of sqlalchemy by the driver should be an implementation detail, not exposed to users. You can wrap a row with `dict()` before returning, or even better use a dataclass/nametuple/etc.
	- The `ranges()` function especially needs improvement: it returns an sqlalchemy row record, for which "bounds" JSONB is returned as a raw string. So I call `bounds()` and then`json.loads(row["bounds"])` to get a dict of top/bottom/left/right fields. (sqlalchemy usually returns jsonb as a dict, but the code is using table reflection rather than your table definitions, so it's not identifying the postgres types properly).
	- Most of our other index apis use a `datacube.model.Range` object for a range, and elsewhere we  sometimes use a two-element tuple. I'd use one of these rather than adding a third representation (separate "time_min" and "time_max" fields)

- The ability to query arbitrary extents is neat! 


A few points I'm unsure about — take with a grain of salt!

- The use of the uuid for extent lookup. Note that it's larger in size than using the fields themselves as a composite (dataset_type_ref is 2 bytes, start date is 4 bytes, offset is char 1 (or 2?) bytes == 8 bytes, vs UUID 16 bytes). It has the downside of being unreadable/obtuse by developers when looking at the table, and is difficult to query/compute in SQL. It presumably may index better though, avoiding a composite index? I haven't tested.

- Storing/using pandas offset aliases as the basis of the api. This seems extremely flexible at the top api level, but I'm less sure about storing them at the low level. It's doing exact string matching on "1M" and "1Y", but in theory users could supply lots of other variations ("1m", "1m0d" ... I think?). These strings are stored in Postgres but can't be understood by Postgres  queries themselves (unless we assume uses only every pass exactly "1M" and "1Y"). If we *effectively* always expect month/days, perhaps we could just use a simple enum? Or if flexibility is needed, we could look at a more portable/convertible representation for it: Postgres supports iso durations and/or has its own duration type.

- The range `bounds` column is defined as `jsonb` but is always four numbers: left/right/up/down. In lieu of adding Postgis as a dependency, the built-in Postgres `box` type or even a four-element array are simpler and much smaller on disk, assuming we're not using the field dynamically(?)

---

Please feel free to come and discuss any of these, especially testing or the drivers/layering!

jeremyh commented:
Also, I'd open the floor for team discussions on the names of things. Damien may be better at this than I am.

- The `db_extent` module name. It's inconsistent with our other modules, and "db_" is unnecessary. Maybe just `extent`, or remove it entirely if it's being merged into the index package.

- Table/entity names:

| Name | Current Python Comment | Maybe rename to? | 
| :--- | :--- | :--- |
| `extent` | "The spatial extent geometry for various time periods. Refer to extent_meta table to obtain length of time period." | `extent_slice` |
| `extent_meta` |  "Metadata for spatial extents." | `extent` |
| `ranges` | "Time min/max and rectangular spatial bounds for products" | `product_range` (or `dataset_type_range` for table consistency: it's probably less confusing until we rename the other one) |

- Some of the api function name could be reviewed. `get_extent()` and `get_extent_direct()` could be more distinguishable from each other?

ashoka1234 committed:
Test set up for PostgresDbAPI extent function tests

ashoka1234 commented:
Thanks Jeremy. Will get onto those one by one after discussing with Damien.

ashoka1234 commented:
The stuff in _extent_upload is not merged yet and we haven't given priority as far as merge with existing code is concerned. The retrieval stuff are now more integrated with exiting code so does not use any driver specific details as far as functions that are merged in ProductResource concerned. Also we no longer use uuid's for retrieval  so I guess it could just be removed altogether or replaced with a int?

ashoka1234 committed:
Refactoring to reflect schema changes, name changes, and more merging of db-extent code into PostgresDbAPI class.
Integration test scripts are cleaned up further in order to use pytest fixtures.

ashoka1234 committed:
All the database retrieval and storage functions of db-extent module are merged with datacube-core.

ashoka1234 committed:
merged extent computes into ComputeResource in index module

ashoka1234 committed:
corrected linting and refactoring errors

ashoka1234 committed:
corrected refactoring errors.

ashoka1234 committed:
Added shapely to setup

ashoka1234 committed:
has_extents_ranges function updates to reflect schema name changes.

omad commented:
We still want a feature like this, but there's too many outstanding issues with this to be able to merge.
